### ISSUE TITLE
Support equivalent of stack `--no-load` with `cabal repl`

### ISSUE THREAD

**Issue by sirlensalot** [haskell/cabal#7541]
... namely, not loading all the modules of the library when kicking off `cabal repl`.

**Comment by emilypi** [comment]
So this one is going to be code changes in one file.

Add the option here: 

- https://github.com/haskell/cabal/blob/1be581d5463adebf8dc8cc8a5375a9ca3051970a/cabal-install/src/Distribution/Client/CmdRepl.hs#L127

And then you'll need to map that arg to a particular functionality here: 

- https://github.com/haskell/cabal/blob/1be581d5463adebf8dc8cc8a5375a9ca3051970a/cabal-install/src/Distribution/Client/CmdRepl.hs#L199

It looks like the current `replAction` loads in the project details by default. Adding `--no-load` will amount to making sure that if that flag is present, then we don't go through with the project load, only bringing up the ghci instance. Beyond that, a changelog entry and doc updates are all that are required.

**Comment by gbaz** [comment]
can't you just accomplish this by `cabal exec ghci`?

**Comment by emilypi** [comment]
Huh that does seem to work, but only after a build: 

```
λ P pact → λ git master → cabal exec ghci
GHCi, version 8.8.4: https://www.haskell.org/ghc/  :? for help
Loaded package environment from /Users/emilypi/kadena/pact/dist-newstyle/tmp/environment.-8869/.ghc.environment.x86_64-darwin-8.8.4
Loaded package environment from /Users/emilypi/kadena/pact/dist-newstyle/tmp/environment.-8869/.ghc.environment.x86_64-darwin-8.8.4
Loaded GHCi configuration from /Users/emilypi/.ghc/ghci.conf
П> :r
Ok, no modules loaded.
П> :q
Leaving GHCi.
λ P pact → λ git master → cabal clean
λ P pact → λ git master → cabal exec ghci
Resolving dependencies...
GHCi, version 8.8.4: https://www.haskell.org/ghc/  :? for help
Loaded package environment from /Users/emilypi/kadena/pact/dist-newstyle/tmp/environment.-8924/.ghc.environment.x86_64-darwin-8.8.4
ghc: can't find a package database at /Users/emilypi/kadena/pact/dist-newstyle/packagedb/ghc-8.8.4
```

Seems like it requires an env file to be present in order to work at all.

**Comment by fendor** [comment]
I believe such a flag requires a bit more changes than just in `CmdRepl`, since the `repl` action is something that `Cabal` performs. As such, what modules are actually loaded. If we execute `cabal repl -v2` in a project, we can see:
```
> cabal repl -v2
...
act-as-setup --build-type=Simple -- repl --verbose=2
...
```
which specifies no targets (due to how configure works), but then eventually:
```
/home/hugin/.ghcup/bin/ghc --interactive -fbuilding-cabal-package <all-these-includes> <package-flags> -XHaskell2010 <list-of-modules>
```
What we would care about is the list of modules, afaict, and these are added by `Cabal`.

**Comment by fgaz** [comment]
> can't you just accomplish this by cabal exec ghci?

or `--write-ghc-environment-files=always`

**Comment by emilypi** [comment]
Two questions about that approach @fgaz: 

- Can you set project targets at will with environment files?
- That setting is off by default for a good reason - it conflicts with nix setups and other flavors of reproduceability, so we probably shouldn't be suggesting people mess around with them as a default. If we can, we should ask `cabal-install` to just... not load targets if requested. Has the wisdom from the past 4 years changed on the subject of env files?  (https://github.com/NixOS/nixpkgs/issues/62906, https://github.com/i-am-tom/dagmore/pull/1, https://www.reddit.com/r/haskell/comments/9a6fg4/package_environment_files_run_counter_to/, https://github.com/tweag/rules_haskell/issues/516)

**Comment by gbaz** [comment]
There's a confusion here -- the `always` flag isn't setting it by _default_ its just setting it _for that project_. And you can always delete the file. Its a good and useful feature, as long as you're aware you're using it (i.e. by explicitly asking for it). That _was_ the wisdom of the past four years. They're a key, important feature. Just don't hide the fact they're in use.

Environment files don't interact with project targets. They just bring deps into scope (through providing a virtual packagedb, in essence). If the test suite has a bigger dep footprint than the lib, then building with tests enabled will create an environment with dependencies in scope to let you load the test suite, the lib, etc.

Again there's two solutions presented here that seem _just fine_ for the requested functionality -- either `cabal exec` (which uses an env file under the hood, iirc) or writing an env file explicitly. Both seem very lightweight. In such a case, I'd advocate we don't add more flags and complexity to the cabal executable itself...

**Comment by enobayram** [comment]
I also need the ability to load a subset of my package's modules in `cabal repl` without loading the entire package first, but I can't follow [the environment files discussion above](https://github.com/haskell/cabal/issues/7541#issuecomment-898937026). I know that having a `--no-load` flag that simply omits the [`list-of-modules`](https://github.com/haskell/cabal/issues/7541#issuecomment-898909563) would solve my issue, and I've already tried `cabal exec ghci`, but the latter is missing all the modules normally generated during a build.

**Comment by gbaz** [comment]
@enobayram what do you mean "missing all the modules normally generated"? I don't understand your comment fully.

**Comment by enobayram** [comment]
@gbaz To be specific, I've tried running `cabal exec ghci -- -imy-src-folder` which drops me into a ghci session where I can `:load` modules from my package, but I'll get errors if I need any of the modules like `My_package_Paths` or those generated by, say `happy` from `.y` parser definitions. `cabal repl` seems to handle such generated modules fine. 

BTW, I've tried running `cabal repl -v`, noted the `ghc -interactive` call made by it and manually removed all the module names from the arguments list. That drops me into the perfect ghci session where nothing is loaded by default, but I can load any module from my package, so it'd make my life so much easier to be able to run:

```
cabal repl --no-load --repl-options "-ghci-script some-script-that-loads-a-module-and-then-runs-some-ghci-commands.ghci"
```

Because as long as `cabal repl` (without `--no-load`) adds those modules, whatever I do with my `-ghci-script` gets clobbered while loading my ghci script.
